File: tmp/fix_cloze.js
const fs = require('fs');
const path = require('path');

const examsDir = 'c:\\Users\\User\\Desktop\\Adai\\yds\\ydsgrammar\\public\\exams\\full';
const files = fs.readdirSync(examsDir).filter(f => f.endsWith('.json'));

function cleanQuestion(text) {
    return text.replace(/\(Cloze Test\s+\d+.*?\)/gi, '')
               .replace(/\(Paragraph\s+\d+.*?\)/gi, '')
               .replace(/\(Passage\s+\d+.*?\)/gi, '')
               .replace(/\[Text as above\]/gi, '')
               .replace(/- Question\s+\d+/gi, '')
               .trim();
}

function extractProperPassage(questions) {
    if (!questions || questions.length === 0) return '';
    
    // Find the longest cleaned question text
    const cleaned = questions.map(q => cleanQuestion(q.question));
    let longest = '';
    for (const t of cleaned) {
        if (t.length > longest.length) longest = t;
    }

    // Check if the longest contains most of the others
    // For Cloze, they often have the full text in one or all.
    // If they were fragments joined before, the "longest" might be the 5x joined one.
    // So we need to detect if it's already a 5x joined mess.
    
    if (longest.includes('- Question')) {
        // It's probably a joined mess. Split it and take one part.
        const parts = longest.split(/- Question\s+\d+/i);
        return parts[0].trim();
    }
    
    // If it has multiple blanks and looks like a full paragraph, it's good.
    // If it's a fragment but we joined it, we need to be careful.
    return longest;
}

files.forEach(file => {
    const filePath = path.join(examsDir, file);
    const data = JSON.parse(fs.readFileSync(filePath, 'utf8'));
    let changed = false;

    // Fix cloze blocks
    const clozeSections = data.sections.filter(s => s.id === 'cloze');
    clozeSections.forEach(sec => {
        for (let i = sec.from; i <= sec.to; i += 5) {
            const block = data.questions.filter(q => q.id >= i && q.id < i + 5);
            if (block.length > 0) {
                const passageId = block[0].passage_id;
                if (passageId) {
                    const passIdx = data.passages.findIndex(p => p.id === passageId);
                    if (passIdx !== -1) {
                        const originalText = data.passages[passIdx].text;
                        // Better extraction logic:
                        // If the text contains "- Question", it's a messed up join.
                        if (originalText.includes('- Question')) {
                            const parts = originalText.split(/- Question\s+\d+/i);
                            data.passages[passIdx].text = parts[0].trim();
                            changed = true;
                        }
                    }
                }
            }
        }
    });

    if (changed) {
        fs.writeFileSync(filePath, JSON.stringify(data, null, 4), 'utf8');
        console.log(`Fixed ${file}`);
    }
});

File: tmp/fix_cloze.py
import os
import json
import re

base_dir = r'c:\Users\User\Desktop\Adai\yds\ydsgrammar\public\exams'
folders = [os.path.join(base_dir, 'full'), os.path.join(base_dir, 'mini')]

for folder in folders:
    if not os.path.exists(folder):
        continue
    for filename in os.listdir(folder):
        if filename.endswith('.json'):
            file_path = os.path.join(folder, filename)
            try:
                with open(file_path, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                
                changed = False
                if 'passages' in data and isinstance(data['passages'], list):
                    for p in data['passages']:
                        if 'text' in p and p['text'] and '- Question' in p['text']:
                            parts = re.split(r'- Question\s+\d+', p['text'], flags=re.IGNORECASE)
                            if len(parts) > 0 and len(parts[0]) > 20:
                                p['text'] = parts[0].strip()
                                changed = True
                
                if 'questions' in data and isinstance(data['questions'], list):
                    for q in data['questions']:
                        if 'question' in q and q['question'] and '- Question' in q['question']:
                            parts = re.split(r'- Question\s+\d+', q['question'], flags=re.IGNORECASE)
                            if len(parts) > 0:
                                q['question'] = parts[0].strip()
                                changed = True
                
                if changed:
                    with open(file_path, 'w', encoding='utf-8') as f:
                        json.dump(data, f, indent=4, ensure_ascii=False)
                    print(f"Fixed {filename}")
            except Exception as e:
                print(f"Error {filename}: {str(e)}")
